src/network/network_gui.cpp

```diff
@@ -1564,333 +1575,735 @@ NetworkCompanyInfo *GetLobbyCompanyInfo(CompanyID company)
 	{
-		/* Check if we need to reset the height */
-		if (!this->CheckClientListHeight()) return;
+		for (auto &button : *buttons) {
+			bool rtl = _current_text_dir == TD_RTL;
 
-		this->DrawWidgets();
+			uint left = rtl ? x : x - button->width;
+			uint right = rtl ? x + button->width : x;
+
+			int offset = std::max(0, ((int)(this->line_height + 1) - (int)button->height) / 2);
+
+			DrawFrameRect(left, y + offset, right, y + offset + button->height, button->colour, FR_NONE);
+			DrawSprite(button->sprite, PAL_NONE, left + ScaleGUITrad(WD_FRAMERECT_LEFT), y + offset + ScaleGUITrad(WD_FRAMERECT_TOP));
+			if (button->disabled) {
+				GfxFillRect(left + 1, y + offset + 1, right - 1, y + offset + button->height - 1, _colour_gradient[button->colour & 0xF][2], FILLRECT_CHECKER);
+			}
+
+			uint width = button->width + WD_FRAMERECT_LEFT + WD_FRAMERECT_RIGHT;
+			x += rtl ? width : -width;
+		}
+
+		return x;
 	}
 
-	void DrawWidget(const Rect &r, int widget) const override
+	uint DrawCompany(CompanyID company_id, uint left, uint right, uint top, uint count) const
 	{
-		if (widget != WID_CL_PANEL) return;
-
 		bool rtl = _current_text_dir == TD_RTL;
-		int icon_offset = (this->line_height - icon_size.height) / 2;
-		int text_offset = (this->line_height - FONT_HEIGHT_NORMAL) / 2;
+		int text_y_offset = std::max(0, ((int)(this->line_height + 1) - (int)FONT_HEIGHT_NORMAL) / 2) + WD_MATRIX_BOTTOM;
 
-		uint y = r.top + WD_FRAMERECT_TOP;
-		uint left = r.left + WD_FRAMERECT_LEFT;
-		uint right = r.right - WD_FRAMERECT_RIGHT;
-		uint type_icon_width = this->server_client_width + this->icon_size.width + WD_FRAMERECT_LEFT;
+		Dimension d = GetSpriteSize(SPR_COMPANY_ICON);
+		int offset = std::max(0, ((int)(this->line_height + 1) - (int)d.height) / 2);
 
+		uint text_left = left + (rtl ? (uint)WD_FRAMERECT_LEFT : d.width + 8);
+		uint text_right = right - (rtl ? d.width + 8 : (uint)WD_FRAMERECT_RIGHT);
 
-		uint type_left  = rtl ? right - this->server_client_width : left;
-		uint type_right = rtl ? right : left + this->server_client_width - 1;
-		uint icon_left  = rtl ? right - type_icon_width + WD_FRAMERECT_LEFT : left + this->server_client_width;
-		uint name_left  = rtl ? left : left + type_icon_width;
-		uint name_right = rtl ? right - type_icon_width : right;
+		uint count_start = this->vscroll->GetPosition();
+		uint count_end = count_start + this->vscroll->GetCapacity();
 
-		int i = 0;
-		for (const NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
-			TextColour colour;
-			if (this->selected_item == i++) { // Selected item, highlight it
-				GfxFillRect(r.left + 1, y, r.right - 1, y + this->line_height - 1, PC_BLACK);
-				colour = TC_WHITE;
-			} else {
-				colour = TC_BLACK;
+		uint y = top + (this->line_height * (count - count_start));
+
+		/* Draw the company line (if in range of scrollbar). */
+		if (count >= count_start && count < count_end) {
```

Comment:
What about `IsInsideMM(count, count_start, count_end)`? Could also be used later in this function.